WRP-472: Migrate project to ESM (#25)
enactjs/jsdoc-to-ts: README.md

## Usage

> Assuming this directory is a peer of the Sandstone source and you want to write this into a same path of sandstone:

Reviewer: Why are you changing enact to sandstone?

Author: An example of enact is as follows. 
node bin/jsdoc-to-ts.js ../enact/core -o=../enact/core
node bin/jsdoc-to-ts.js ../enact/ui-o=../enact/ui
node bin/jsdoc-to-ts.js ../enact/i18n -o=../enact/i18n 
node bin/jsdoc-to-ts.js ../enact/webos -o=../enact/webos 
node bin/jsdoc-to-ts.js ../enact/spotlight -o=../enact/spotlight 

Sandstone looks good to show the simplicity of the CLI method.

```bash

> node bin/jsdoc-to-ts.js ../sandstone -o=../sandstone

Reviewer: 🤔 I'm confused. I think there was a reason to document this before.
The content is the same one as bin/jsdoc-to-ts. Would you please figure out just to replace `require` to `import` or so?

Author: - I think, `node -e "script"` is the initial method, and since the #12 changes, it has been changed to the CLI method, but it does not appear to be used in the user guide (because it still provides the initial method).
- I searched the way but cannot find... I think import() cannot be used directly in the Bash command line because it is a JavaScript feature used in ESM and is not recognized by Bash. I am concerned that If we should migrate the ESM style, then we need to drop `node -e "script"` usage.

## Usage

Assuming this directory is a peer of the Enact source and you want to write this into an installed (from npm) version of enact:

> (This script requires at least Node.js 13.2.0.)

Reviewer: Let's move this to Note. at line 36.

Author: OK

```bash
node -e "import('./index.js').then(({default: jsdocToTs}) => {
  ['core', 'ui', 'moonstone', 'i18n', 'webos', 'spotlight'].forEach(p => jsdocToTs({
    package: '../enact/packages/' + p,
    output: require('fs').writeFileSync,
    ignore: ['node_modules', 'ilib'],
    importMap: {
      ui: '@enact/ui',
      moonstone: '@enact/moonstone',
      core: '@enact/core',
      webos: '@enact/webos',
      spotlight: '@enact/spotlight',
      i18n: '@enact/i18n'
    },
    outputPath: '<path to installation>/node_modules/@enact/' + p
  }))
})"
```
NOTE: Replace `<path to installation>` with the directory you wish to update.

## Working with Linked Enact

The following command will parse the Enact source into a local `types` directory so you can use TypeScript with linked Enact:

> (This script requires at least Node.js 13.2.0.)

Reviewer: I don't think we need to add this here.

Author: OK

> ```bash

Reviewer: ```suggestion

```bash
```

Author: Done

Assuming this directory is a peer of the Enact source and you want to write this into an installed (from npm) version of enact:

```bash

> node -e "import('./index.js').then((jsdocToTs) => {

Reviewer: Try
...then({default: jsdocToTs}) => {

Reviewer: And call jsdocToTs({...

Author: OK

> node -e "import('./index.js').then(({default: jsdocToTs}) => {

Reviewer: I think we need to mention using `import` from which version of the node is possible.
What do you think?

Author: You're right. Let's add the required node version.

  ['core', 'ui', 'moonstone', 'i18n', 'webos', 'spotlight'].forEach(p => jsdocToTs({
    package: '../enact/packages/' + p,
    output: require('fs').writeFileSync,
    ignore: ['node_modules', 'ilib'],
    importMap: {
      ui: '@enact/ui',
      moonstone: '@enact/moonstone',
      core: '@enact/core',
      webos: '@enact/webos',
      spotlight: '@enact/spotlight',
      i18n: '@enact/i18n'
    },
    outputPath: '<path to installation>/node_modules/@enact/' + p
  }))
})"
```
NOTE: Replace `<path to installation>` with the directory you wish to update.

## Working with Linked Enact

The following command will parse the Enact source into a local `types` directory so you can use TypeScript with linked Enact:
```bash
node -e "import('./index.js').then(({default: jsdocToTs}) => {

>     ['core', 'ui', 'moonstone', 'i18n', 'webos', 'spotlight'].forEach(p => jsdocToTs('.')({

Reviewer: I don't think this command works. Is this right?

Author: There's an error in the guide. I will fix it.

Assuming this directory is a peer of the Enact source and you want to write this into an installed (from npm) version of enact:

```bash
node -e "import('./index.js').then(({default: jsdocToTs}) => {
  ['core', 'ui', 'moonstone', 'i18n', 'webos', 'spotlight'].forEach(p => jsdocToTs({
    package: '../enact/packages/' + p,
    output: require('fs').writeFileSync,
    ignore: ['node_modules', 'ilib'],
    importMap: {
      ui: '@enact/ui',
      moonstone: '@enact/moonstone',
      core: '@enact/core',
      webos: '@enact/webos',
      spotlight: '@enact/spotlight',
      i18n: '@enact/i18n'
    },
    outputPath: '<path to installation>/node_modules/@enact/' + p
  }))
})"
```

> NOTE: Replace `<path to installation>` with the directory you wish to update (This script requires at least Node.js 13.2.0.).

Reviewer: ```suggestion
NOTE: Replace `<path to installation>` with the directory you wish to update. The script requires at least Node.js 13.2.0. or later.
```

## Working with Linked Enact

The following command will parse the Enact source into a local `types` directory so you can use TypeScript with linked Enact:

```bash
node -e "import('./index.js').then(({default: jsdocToTs}) => {
    ['core', 'ui', 'moonstone', 'i18n', 'webos', 'spotlight'].forEach(p => jsdocToTs({
    package: '<path to enact>/packages/' + p,
    output: (filepath, content) => {
      const path = require('path');
      const dir = path.dirname(filepath);
      const wfs = require('fs').writeFileSync;
      // recursively create the path to the folder in types
      require('mkdirp').sync(dir);
      // write a package.json with a types member pointing to the .d.ts file
      wfs(path.join(dir, 'package.json'), '{\"types\": \"' + path.basename(filepath) + '\"}');
      // write the .d.ts file
      wfs(filepath, content);
    },
    importMap: {
      ui: '@enact/ui',
      moonstone: '@enact/moonstone',
      core: '@enact/core',
      webos: '@enact/webos',
      spotlight: '@enact/spotlight',
      i18n: '@enact/i18n'
    },
    outputPath: '<path to installation>/types/@enact/' + p
  }))
})"
```

> NOTE: You need to install mkdirp module in jsdoc-to-ts by yourself.

Reviewer: ```suggestion
NOTE: You need to install `mkdirp` module in jsdoc-to-ts by yourself.
```
